src/core/evaluator.js

```diff
@@ -1037,8 +1037,66 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
         } else {
           textChunk.height += Math.abs(height * scaleCtmX * scaleLineX);
         }
+        addSpaceIfNecessary(textChunk, font);
         return textChunk;
       }
+      function getChunkPos(chunk, font) {
+        var tx = chunk.transform;
+        var angle = Math.atan2(tx[1], tx[0]);
+        if (font.vertical) {
+          angle += Math.PI / 2;
+        }
+        var fontHeight = Math.sqrt((tx[2] * tx[2]) + (tx[3] * tx[3]));
+        var fontAscent = fontHeight;
+        if (font.ascent) {
+          fontAscent = font.ascent * fontAscent;
+        } else if (font.descent) {
+          fontAscent = (1 + font.descent) * fontAscent;
+        }
+        return {
+          x : (angle === 0 ? tx[4] : tx[4] + (fontAscent * Math.sin(angle))),
+          y: tx[5]
+        };
+      }
+      function addSpaceIfNecessary(newChunk, font) {
+        // If the new chunk starts with a space, it does not need one.
+        if (newChunk.str[0] === ' ' || newChunk.str[0] === '-') {
+          return;
+        }
+        if (bidiTexts.length === 0) {
+          return;
+        }
+        // If the last chunk ends with a space it does not need one.
+        var lastChunk = bidiTexts[bidiTexts.length - 1];
+        if (lastChunk.str.length === 0) {
+          return;
+        }
+        var lastChar = lastChunk.str[lastChunk.str.length - 1];
+        if (lastChar === ' ' || lastChar === '-') {
+          return;
+        }
+        var lastPos = getChunkPos(lastChunk, font);
+        var newPos = getChunkPos(newChunk, font);
+        var yDiff = Math.abs(lastPos.y  - newPos.y);
+        if (yDiff >= lastChunk.height || yDiff >= newChunk.height) {
+          // On different lines, add a space.
+          lastChunk.str += ' ';
+        } else {
+          var wordSpacing = textState.wordSpacing > 0 ?
```

Comment:
Let's use a regular `if` statement here instead of the ternary operator for readability. I think the ternary operator should only be used if the statement can (reasonably) be put on a single line; this statement including the comments is too long for that. Also applies to `addSpace` below.